### PR TITLE
Allocate HIR on an arena 2/4 -- Expr & Pat

### PR DESCRIPTION
**Pull request by cjgillot**
This is the second PR in the series started by #66931 

This time, commits don't really make sense on their own.
They are mostly split by type of compile error.

The additional diff is here: https://github.com/cjgillot/rust/compare/hirene-preamble...hirene-expr

### CODE REVIEW

**Comment by davidtwco** [comment]
r? @eddyb 
cc @Zoxc

**Comment by cjgillot** [comment]
I seem to be constructing some sort of infinite loop in the HIR graph. The stack overflow happens when hashing everything. The loop consists of a `ExprKind::Call(ExprKind::Path(QPath::TypeRelative(...)), &[ExprKind::Binary(..., loop, ...))`.

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -51,6 +51,15 @@ macro_rules! hir_vec {
     );
 }
 
+macro_rules! arena_vec {
```
Comment (Zoxc) [inline src/librustc/hir/mod.rs]: This should be in `lowering.rs`.

**Comment by Zoxc** [comment]
Did you ensure that the `Fewer calls to arena.alloc` commit didn't result in allocations which wouldn't get used?

And did you resolve the infinite loop?

File: src/librustc/hir/mod.rs (hunk 1 above)
Comment (cjgillot) [reply src/librustc/hir/mod.rs]: I did not manage to get it imported in the lowering sub-modules. I do not know this corner of the language enough.

**Comment by cjgillot** [comment]
Rebased.

> Did you ensure that the `Fewer calls to arena.alloc` commit didn't result in allocations which wouldn't get used?
> 

I did as much as I could. There are no unused allocation of HIR nodes, because of typing. However, typing does not guarantee against unused allocation of a pointer.

> And did you resolve the infinite loop?

Yes. It was in the arena allocation code. Bug #67001

File: src/librustc/hir/mod.rs (hunk 1 above)
Comment (Centril) [reply src/librustc/hir/mod.rs]: Should be fine, we can move it later.

Comment (Zoxc) [reply src/librustc/hir/mod.rs]: The macro definition has to be before the declaration of the modules in `lowering.rs`.

**Comment by Zoxc** [comment]
> I did as much as I could.

As in you looked for cases where unused allocation could happen but did not find any?

> There are no unused allocation of HIR nodes, because of typing. However, typing does not guarantee against unused allocation of a pointer.

I'm not sure what you're saying here.

**Comment by cjgillot** [comment]
> As in you looked for cases where unused allocation could happen but did not find any?

Yes.

> I'm not sure what you're saying here.

I mean typing would prevent an allocation `&'hir Expr` from being leaked, but may allow an allocation `&'hir &'hir Expr` to be leaked. This should not represent a significant amount of memory.

**Comment by cjgillot** [comment]
Rebased.

**Comment by cjgillot** [comment]
Rebased with review comments.

**Comment by cjgillot** [comment]
Rebased.

**Pull request merged**